File: backend/internal/api/router.go
package api

import (
	"github.com/Meheer17/ambieye/internal/api/handlers"
	"github.com/Meheer17/ambieye/internal/api/middleware"
	"github.com/Meheer17/ambieye/internal/config"
	"github.com/gin-contrib/cors"
	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/mongo"
)

// SetupRouter configures the API routes
func SetupRouter(client *mongo.Client, cfg *config.Config) *gin.Engine {
	router := gin.Default()

	// Configure CORS
	corsConfig := cors.DefaultConfig()
	corsConfig.AllowOrigins = []string{"*"} // In production, specify your actual domains
	corsConfig.AllowMethods = []string{"GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"}
	corsConfig.AllowHeaders = []string{"Origin", "Content-Type", "Accept", "Authorization"}
	router.Use(cors.New(corsConfig))

	// Get database
	db := client.Database(cfg.MongoDB.Database)

	// Initialize handlers
	authHandler := handlers.NewAuthHandler(db, cfg)
	notificationHandler := handlers.NewNotificationHandler(db)
	doctorHandler := handlers.NewDoctorHandler(db)
	patientHandler := handlers.NewPatientHandler(db)
	queryHandler := handlers.NewQueryHandler(db, notificationHandler)

	// Authentication middleware
	authMiddleware := middleware.AuthMiddleware(cfg.JWT.Secret)
	doctorMiddleware := middleware.RoleMiddleware("doctor")
	patientMiddleware := middleware.RoleMiddleware("patient")

	router.GET("/privacy-policy", func(c *gin.Context) {
		c.File("./public/privacy.html")
	})
	router.GET("/delete-account", func(c *gin.Context) {
		c.File("./public/delete.html")
	})

	// API routes
	api := router.Group("/api")
	{
		// Auth routes - no auth required
		auth := api.Group("/auth")
		{
			auth.POST("/login", authHandler.Login)
			auth.POST("/signup", authHandler.Signup)
			auth.POST("/logout", authHandler.Logout) // Client-side token removal
			auth.GET("/verify", authMiddleware, authHandler.Verify)
		}

		// Doctor routes - require auth and doctor role
		doctor := api.Group("/doctor", authMiddleware, doctorMiddleware)
		{
			doctor.GET("/dashboard", doctorHandler.GetDashboard)
			doctor.GET("/patients", doctorHandler.GetPatients)
			doctor.GET("/patients/:id", doctorHandler.GetPatientById)
			doctor.POST("/patients/medicalinfo/:id", doctorHandler.UpdatePatientMedicalInfo)
			doctor.POST("/patients/visitrecord/:id", doctorHandler.AddPatientVisitRecord)
			doctor.GET("/profile", doctorHandler.GetProfile)
			doctor.GET("/queries", queryHandler.GetAllQueries)
			doctor.POST("/delete", doctorHandler.DeleteDoctor)
		}

		// Patient routes - require auth and patient role
		patient := api.Group("/patient", authMiddleware)
		{
			patient.GET("/dashboard", patientHandler.GetDashboard)
			patient.GET("/profile", patientHandler.GetProfile)
			patient.GET("/doctors/:id", patientMiddleware, patientHandler.GetDoctorById)
			patient.PUT("/profile", patientHandler.UpdateProfile)
			patient.GET("/queries", patientHandler.GetPatientQueries)
			patient.POST("/queries", patientMiddleware, queryHandler.CreateQuery)
			patient.POST("/delete", patientHandler.DeletePatient)
		}

		gameHandler := handlers.NewGameHandler(db)
		games := api.Group("/games", authMiddleware)
		{
			games.POST("/results", patientMiddleware, gameHandler.SaveGameResult)
			games.GET("/today", gameHandler.GetTodayGameResults)
			games.GET("/history", gameHandler.GetGameHistory)
		}

		// Query routes - require auth, role checked in handlers
		queries := api.Group("/queries", authMiddleware)
		{
			queries.GET("/:id", queryHandler.GetQuery)
			queries.PUT("/:id/answer", queryHandler.AnswerQuery) // Doctor only
		}

		// Notification routes - require auth
		// notifications := api.Group("/notifications", authMiddleware)
		// {
		// 	notifications.POST("/register", notificationHandler.RegisterDevice)
		// 	notifications.GET("/", notificationHandler.GetNotifications)
		// 	notifications.GET("/settings", notificationHandler.GetNotificationSettings)
		// 	notifications.PUT("/settings", notificationHandler.UpdateNotificationSettings)
		// }
	}

	return router
}

File: backend-php/src/api/router.php
<?php

namespace Api;

use Config\Config;
use Api\Handlers\AuthHandler;
use Api\Handlers\DoctorHandler;
use Api\Handlers\PatientHandler;
use Api\Handlers\QueryHandler;
use Api\Handlers\GameHandler;
use Api\Handlers\NotificationHandler;
use Api\Middleware\AuthMiddleware;
use Api\Middleware\RoleMiddleware;

class Router {
    private $config;
    private $method;
    private $path;
    private $handlers;

    public function __construct(Config $config) {
        $this->config = $config;
        $this->method = $_SERVER['REQUEST_METHOD'];
        $this->path = parse_url($_SERVER['REQUEST_URI'], PHP_URL_PATH);
        
        // Initialize handlers
        $this->handlers = [
            'auth' => new AuthHandler($config),
            'doctor' => new DoctorHandler(),
            'patient' => new PatientHandler(),
            'query' => new QueryHandler(),
            'game' => new GameHandler(),
            'notification' => new NotificationHandler(),
        ];
    }

    public function route() {
        // Extract the path from REQUEST_URI
        $this->path = parse_url($_SERVER['REQUEST_URI'], PHP_URL_PATH);
        
        // Remove base paths - handle various server configurations
        $this->path = preg_replace('#^/Ambieye/backend-php/public#', '', $this->path);
        $this->path = preg_replace('#^/public#', '', $this->path);
        
        // Ensure path starts with /
        $this->path = '/' . ltrim($this->path, '/');
        // Remove trailing slash but keep root as /
        $this->path = rtrim($this->path, '/') ?: '/';

        // Set CORS headers
        $this->setCorsHeaders();

        // Handle preflight requests
        if ($this->method === 'OPTIONS') {
            http_response_code(200);
            exit;
        }

        // Set JSON header
        header('Content-Type: application/json');

        // Parse JSON body
        $body = json_decode(file_get_contents('php://input'), true) ?? [];

        // Route requests
        switch (true) {
            // Health check
            case $this->path === '/health' && $this->method === 'GET':
                $this->response([
                    'status' => 200,
                    'response' => ['message' => 'API is healthy', 'timestamp' => date('Y-m-d H:i:s')]
                ]);
                break;

            // Static pages
            case $this->path === '/privacy-policy' && $this->method === 'GET':
                $this->serveFile(__DIR__ . '/../../public/privacy.html');
                break;
            case $this->path === '/delete-account' && $this->method === 'GET':
                $this->serveFile(__DIR__ . '/../../public/delete.html');
                break;

            // Auth routes
            case preg_match('#^/api/auth/login$#', $this->path) && $this->method === 'POST':
                $this->response($this->handlers['auth']->login($body));
                break;
            case preg_match('#^/api/auth/signup$#', $this->path) && $this->method === 'POST':
                $this->response($this->handlers['auth']->signup($body));
                break;
            case preg_match('#^/api/auth/logout$#', $this->path) && $this->method === 'POST':
                $this->response($this->handlers['auth']->logout());
                break;
            case preg_match('#^/api/auth/verify$#', $this->path) && $this->method === 'GET':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                $this->response($this->handlers['auth']->verify());
                break;

            // Doctor routes
            case preg_match('#^/api/doctor/dashboard$#', $this->path) && $this->method === 'GET':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                RoleMiddleware::checkRole('doctor');
                $this->response($this->handlers['doctor']->getDashboard());
                break;
            case preg_match('#^/api/doctor/patients$#', $this->path) && $this->method === 'GET':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                RoleMiddleware::checkRole('doctor');
                $this->response($this->handlers['doctor']->getPatients());
                break;
            case preg_match('#^/api/doctor/patients/(\d+)$#', $this->path, $matches) && $this->method === 'GET':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                RoleMiddleware::checkRole('doctor');
                $this->response($this->handlers['doctor']->getPatientById($matches[1]));
                break;
            case preg_match('#^/api/doctor/patients/medicalinfo/(\d+)$#', $this->path, $matches) && $this->method === 'POST':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                RoleMiddleware::checkRole('doctor');
                $this->response($this->handlers['doctor']->updatePatientMedicalInfo($matches[1], $body));
                break;
            case preg_match('#^/api/doctor/patients/visitrecord/(\d+)$#', $this->path, $matches) && $this->method === 'POST':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                RoleMiddleware::checkRole('doctor');
                $this->response($this->handlers['doctor']->addPatientVisitRecord($matches[1], $body));
                break;
            case preg_match('#^/api/doctor/profile$#', $this->path) && $this->method === 'GET':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                RoleMiddleware::checkRole('doctor');
                $this->response($this->handlers['doctor']->getProfile());
                break;
            case preg_match('#^/api/doctor/delete$#', $this->path) && $this->method === 'POST':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                RoleMiddleware::checkRole('doctor');
                $this->response($this->handlers['doctor']->deleteDoctor());
                break;

            // Patient routes
            case preg_match('#^/api/patient/dashboard$#', $this->path) && $this->method === 'GET':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                RoleMiddleware::checkRole('patient');
                $this->response($this->handlers['patient']->getDashboard());
                break;
            case preg_match('#^/api/patient/profile$#', $this->path) && $this->method === 'GET':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                RoleMiddleware::checkRole('patient');
                $this->response($this->handlers['patient']->getProfile());
                break;
            case preg_match('#^/api/patient/profile$#', $this->path) && $this->method === 'PUT':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                RoleMiddleware::checkRole('patient');
                $this->response($this->handlers['patient']->updateProfile($body));
                break;
            case preg_match('#^/api/patient/doctors/(\d+)$#', $this->path, $matches) && $this->method === 'GET':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                RoleMiddleware::checkRole('patient');
                $this->response($this->handlers['patient']->getDoctorById($matches[1]));
                break;
            case preg_match('#^/api/patient/queries$#', $this->path) && $this->method === 'GET':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                RoleMiddleware::checkRole('patient');
                $this->response($this->handlers['patient']->getPatientQueries());
                break;
            case preg_match('#^/api/patient/delete$#', $this->path) && $this->method === 'POST':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                RoleMiddleware::checkRole('patient');
                $this->response($this->handlers['patient']->deletePatient());
                break;

            // Query routes
            case preg_match('#^/api/queries$#', $this->path) && $this->method === 'POST':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                RoleMiddleware::checkRole('patient');
                $this->response($this->handlers['query']->createQuery($body));
                break;
            case preg_match('#^/api/queries/(\d+)$#', $this->path, $matches) && $this->method === 'GET':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                $this->response($this->handlers['query']->getQuery($matches[1]));
                break;
            case preg_match('#^/api/queries/(\d+)/answer$#', $this->path, $matches) && $this->method === 'PUT':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                RoleMiddleware::checkRole('doctor');
                $this->response($this->handlers['query']->answerQuery($matches[1], $body));
                break;
            case preg_match('#^/api/doctor/queries$#', $this->path) && $this->method === 'GET':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                RoleMiddleware::checkRole('doctor');
                $this->response($this->handlers['query']->getAllQueries());
                break;

            // Game routes
            case preg_match('#^/api/games/results$#', $this->path) && $this->method === 'POST':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                RoleMiddleware::checkRole('patient');
                $this->response($this->handlers['game']->saveGameResult($body));
                break;
            case preg_match('#^/api/games/today$#', $this->path) && $this->method === 'GET':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                $this->response($this->handlers['game']->getTodayGameResults());
                break;
            case preg_match('#^/api/games/history$#', $this->path) && $this->method === 'GET':
                $authMiddleware = new AuthMiddleware($this->config->jwt['secret']);
                $authMiddleware->verify();
                $this->response($this->handlers['game']->getGameHistory());
                break;

            default:
                http_response_code(404);
                echo json_encode(['error' => 'Not Found']);
                break;
        }
    }

    private function setCorsHeaders() {
        header('Access-Control-Allow-Origin: *');
        header('Access-Control-Allow-Methods: GET, POST, PUT, PATCH, DELETE, OPTIONS');
        header('Access-Control-Allow-Headers: Origin, Content-Type, Accept, Authorization');
    }

    private function serveFile($filePath) {
        if (file_exists($filePath)) {
            readfile($filePath);
            exit;
        }
        http_response_code(404);
        echo json_encode(['error' => 'File not found']);
    }

    private function response($result) {
        http_response_code($result['status']);
        echo json_encode($result['response']);
        exit;
    }
}
